compiler/src/iree/compiler/Dialect/VM/Transforms/Conversion.cpp

```diff
@@ -128,11 +128,16 @@ class ConversionPass
 
     // MemRef to Util (to VM) is an A->B->C lowering. We must instruct it
     // specifically on what the correct C buffer type is.
-    auto utilBufferType =
+    auto convertedBufferType =
         typeConverter.convertType(IREE::Util::BufferType::get(&getContext()));
-    assert(utilBufferType);
-    populateMemRefToUtilPatterns(context, conversionTarget, typeConverter,
-                                 patterns, utilBufferType);
+    assert(convertedBufferType);
+    auto convertedIndexType =
+        typeConverter.convertType(IndexType::get(&getContext()));
+    assert(convertedIndexType && convertedIndexType.isa<IntegerType>());
+    populateMemRefToUtilPatterns(
```

Comment:
hmmm this is going to be an issue - we need to remove memrefs before lowering into vm as a secondary stage - so linalg->memref->util.buffer ->> vm - as otherwise we are still relying on memrefs even if we don't lower to vm or if we do other things first. Specifically the inline HAL execution work lowers vmvx down using the vmvx transformation pipeline, does a whole bunch of other stuff (inlining into the parent module, applying IPO, then applying HAL lowerings, etc), and only at the very end lowers to vm. If we had memrefs all the way through that we'll have issues. I think I'm ok with the patterns being here until I can make the changes for split transformation but if we require that these be here by relying on VM details then it won't be possible to fix.

I'm not sure when memref<index> makes sense once at this level - we don't support them in tensor form on the outside of the dispatch and I'm not sure how they would arise on the inside. Can we eliminate whatever is introducing them instead? I can't think of valid use cases OTOH and if we have some of these cropping up we're going to have some major issues elsewhere.

Reply:
These are coming from the decomposition of tensor<index>, which is legal. It is typically fused at the higher level but in the process of tiling and decomposing, we end up with them. Then they just get transformed further from there and someone needs to deal with them (i.e. fixate the size of the `index` type). It will be quite hard to eliminate them where they are as an intermediate artifact (i.e. there are ops that have to work on index and then you are in the regime of casts everywhere that need to cancel out) -- they are legal in the conception of these types so there is nothing keeping them from emerging as a result of transformations (within the DR).

Here is the background: https://discourse.llvm.org/t/allowing-memref-of-index-elements/1491

As we discussed offline, the more composable way forward is via a util.sizeof so that the higher level IR can be made generic to target specific sized types.

I did make these parts of the PR a dedicated commit as I expected it to be controversial. Two paths forward:

* Commit as-is and then introduce a util.sizeof cleanup. I can work on that next but would also like to prefetch some next work which is somewhat dependent on the lowerings not being broken (which they were prior to this PR).
* Revert the current index changes and disable the corresponding tests. Followup with reworking through a util.sizeof.

Let me know what you prefer.

Reply:
I think if we moved this to the end of the VMVX transformation pipeline and had it hardcoded to i32 that'd make me ok with it landing with a TODO to remove the hardcoding. My largest concern is that we want the end of VMVX lowering to be util.buffer and if we couple things during vm conversion here that's going to be much harder to split unless the follow-up is happening soon. We could live with the hardcoding for awhile I suspect.